x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java

```diff
@@ -694,62 +694,66 @@ public List<NamedExpression> visitGrouping(EsqlBaseParser.FieldsContext ctx) {
+        if (param == null) {
+            return Literal.NULL;
         }
-        // otherwise we need to make sure that xcontent-serialized value is converted to the correct type
-        try {
+        return visitParam(ctx, param);
+    }
 
-            if (EsqlDataTypeConverter.canConvert(sourceType, dataType) == false) {
-                throw new ParsingException(
-                    source,
-                    "Cannot cast value [{}] of type [{}] to parameter type [{}]",
-                    param.value,
-                    sourceType,
-                    dataType
-                );
-            }
-            return new Literal(source, EsqlDataTypeConverter.converterFor(sourceType, dataType).convert(param.value), dataType);
-        } catch (QlIllegalArgumentException ex) {
-            throw new ParsingException(ex, source, "Unexpected actual parameter type [{}] for type [{}]", sourceType, param.type);
+    private Object visitParam(EsqlBaseParser.ParamsContext ctx, QueryParam param) {
+        Source source = source(ctx);
+        DataType type = param.type();
+        if (type == null) {
+            throw new ParsingException(source, "Invalid parameter data type [{}]", type);
         }
+        return new Literal(source, param.value(), type);
     }
 
-    private TypedParamValue param(TerminalNode node) {
+    QueryParam paramByToken(TerminalNode node) {
         if (node == null) {
             return null;
         }
-
         Token token = node.getSymbol();
-
-        if (params.containsKey(token) == false) {
+        if (params.contains(token) == false) {
             throw new ParsingException(source(node), "Unexpected parameter");
         }
-
         return params.get(token);
     }
 
+    QueryParam paramByNameOrPosition(TerminalNode node) {
+        if (node == null) {
+            return null;
+        }
+        Token token = node.getSymbol();
+        String nameOrPosition = token.getText().substring(1);
+        if (isInteger(nameOrPosition)) {
+            int index = Integer.parseInt(nameOrPosition);
+            if (params.get(index) == null) {
+                String message = "";
+                int np = params.positionalParams().size();
+                if (np > 0) {
+                    message = ", did you mean position " + (np == 1 ? "1?" : " or any between 1 and " + np + "?");
+                }
+                params.addParsingErrors(new Failure(null, "No parameter is defined for position " + index + message));
+            }
+            return params.get(index);
+        } else {
+            if (params.contains(nameOrPosition) == false) {
+                String message = "";
+                Object[] names = params.namedParams().keySet().stream().toArray();
+                if (names.length > 0) {
+                    message = ", did you mean " + (names.length == 1 ? "[" + names[0] + "]?" : " or any of [" + names + " ]?");
```

Comment:
Using `names` here directly in the String won't generate the right comma separated list of values as one might expect. Plus, we need a test for this.

Reply:
Also, ` or ` is the message is not needed.

Reply:
Tests are added and messages are updated.